cpp/src/arrow/buffer.h

```diff
@@ -57,18 +57,25 @@ class ARROW_EXPORT Buffer {
   ///
   /// \note The passed memory must be kept alive through some other means
   Buffer(const uint8_t* data, int64_t size)
-      : is_mutable_(false), is_cpu_(true), data_(data), size_(size), capacity_(size) {
+      : is_mutable_(false), is_cpu_(true), data_(data), size_(size), capacity_(size), device_type_(DeviceType::CPU) {
```

Comment:
The `is_cpu_` boolean is still useful because of pinned host memory. Both `CUDA_HOST` and `ROCM_HOST` are pinned CPU memory which would have the corresponding device type (since the memory is managed by the device drivers) but is accessible to the CPU.

Reply:
While the pointers are accessible to the CPU, there's no guarantee that there isn't stream ordered work happening on them where it could cause a race condition to CPU code accessing the buffer. Given `is_cpu_` is existing ABI, I'm wondering if it wouldn't be a safer option to have it return `False` in these cases where if someone wants to do CPU things against `CUDA_HOST` or `ROCM_HOST` buffers they should explicitly check the device type instead?

Reply:
Hmm, I think the intent for the `is_cpu_` flag was solely to know whether or not the data is accessible via CPU rather than caring about the stream ordered work synchronizations. I intend on expanding the existing classes/objects to make it easier to synchronize for stream ordered work as a separate abstraction, so I lean towards leaving it marked as `true` in this case, but I'm open to changing it.

@pitrou @felipeblazing @Rhonda85 what do you all think?